zerver/lib/bugdown/api_return_values_table_generator.py

```diff
@@ -0,0 +1,71 @@
+import re
+
+from markdown.extensions import Extension
+from markdown.preprocessors import Preprocessor
+from zerver.openapi.openapi import get_openapi_return_values
+from typing import Any, Dict, Optional, List
+import markdown
+
+REGEXP = re.compile(r'\{generate_return_values_table\|\s*(.+?)\s*\|\s*(.+)\s*\}')
+
+
+class MarkdownReturnValuesTableGenerator(Extension):
+    def __init__(self, configs: Optional[Dict[str, Any]]=None) -> None:
+        self.config: Dict[str, Any] = {}
+
+    def extendMarkdown(self, md: markdown.Markdown, md_globals: Dict[str, Any]) -> None:
+        md.preprocessors.add(
+            'generate_return_values', APIReturnValuesTablePreprocessor(md, self.getConfigs()), '_begin'
+        )
+
+
+class APIReturnValuesTablePreprocessor(Preprocessor):
+    def __init__(self, md: markdown.Markdown, config: Dict[str, Any]) -> None:
+        super().__init__(md)
+
+    def run(self, lines: List[str]) -> List[str]:
+        done = False
+        while not done:
+            for line in lines:
+                loc = lines.index(line)
+                match = REGEXP.search(line)
+
+                if not match:
+                    continue
+
+                doc_name = match.group(2)
+                endpoint, method = doc_name.rsplit(':', 1)
+                return_values: Dict[str, Any] = {}
+                return_values = get_openapi_return_values(endpoint, method)
+                text = self.render_table(return_values, 0)
+                line_split = REGEXP.split(line, maxsplit=0)
+                preceding = line_split[0]
+                following = line_split[-1]
+                text = [preceding] + text + [following]
+                lines = lines[:loc] + text + lines[loc+1:]
+                break
+            else:
+                done = True
+        return lines
+
+    def render_table(self, return_values: Dict[str, Any], spacing: int) -> List[str]:
+        obv = ["result", "msg"]
+        ans = []
+        for return_value in return_values:
+            if return_value in obv:
+                continue
+            description = return_values[return_value]['description']
+            description = description.replace('\n', '\n' + ((spacing + 4) * ' '))
+            ans.append((spacing * " ") + "* `" + return_value + "`: " + description)
+            if 'properties' in return_values[return_value]:
+                ans += self.render_table(return_values[return_value]['properties'], spacing + 4)
+            if 'additionalProperties' in return_values[return_value]:
+                ans += self.render_table(return_values[return_value]['additionalProperties']['properties'],
+                                         spacing + 4)
+            if ('items' in return_values[return_value] and
+               'properties' in return_values[return_value]['items']):
+                ans += self.render_table(return_values[return_value]['items']['properties'], spacing + 4)
+        return ans
```

Comment:
This is fine for initial formatting, but we'll likely want to do nicer styling than the current bulleted list format in a future version.  Best to get something merged that lets us migrate the data before getting distracted with that, though.

@akashaviator @eeshangarg can you also review this block of code?